### ISSUE TITLE
access jsPanel object after restoreId

### ISSUE THREAD

**Issue by balage1551** [Flyer53/jsPanel4#146]
This is a great library! 

I stucked with some issues most probably based on my still shallow understanding of the working of the lib.
I use this with Vue and happily integrated. I mostly have fixed sized windows, so I disabled resizing. Also I save the state by the `layout` extension. It works great, with some minor problems.

1. When I first use a new window and not persisted, the layout.restoreId doesn't create it by the options passed. I had to make some hacking:

```
            if (jsPanel.layout.getDataset(this.panelId, 'id', this.storageName)) {
                 jsPanel.layout.restoreId({id: this.panelId, config: options, storagename: this.storageName})
            } else {
                jsPanel.create(options)
            }
```

This is not nice and I am not sure this is the intendted way of operation.

2. In earlier versions the restoreId returned the jsPanel object, but the new one doesn't, as far as I see in the source. How can I aquire a reference of the restored panel?

3. There is no function to remove a saved state from storage. This may be useful.

Thanks for the help!

Balage

**Comment by Flyer53** [comment]
Hi there ... and thanks for the praise 😄 

Your nr 2: I obviously missed that point and fixed it in a preliminary 4.11.3-alpha you can [download](https://alpha.jspanel.de/downloads/jspanel-4.11.3-alpha.zip). `jsPanel.layout.restoreId()` now returns the restored panel. Should be ok if you just swap the layout extension js file.

Your nr 1: I have to check that out and investigate ...

Your nr 3:  So far I didn't see a need for such a method and nobody requested yet. Let me think about this a bit. 

Regards,
Stefan

**Comment by balage1551** [comment]
Thanks for the quick answer!
I am getting used to the framework and find it more and more awesome.

About my questions. Thanks the quick fix. I'll wait for the next release with the fix of Question 1 the my code would be much nicer.
About the 3rd, I give you two scenarios:
The first: the size and/or position of the panel changes by a new release and I would like to follow it. Now I have to code a hack to manually call resize/setPosition to override the saved size. 
The second: the user of my app made a mistake (because I forget to set the containment and dragged the panel out of the screen on the top. There is no easy way to recover from this: you have to either code a repositioning into your code to help him, or he has to find the option in the browser to clear local storage of the file. Not an easy task. If I could reset the state, my code become an one-liner by adding a reset layout option.
Third reason: I would like to test how my app will behave when first opened by a user and there is no persisted state. Now I have to manually kill the storage.

I could add to my code: `localStorage.removeItem("jsPanels")`, but this clears the whole config. 

Regards,
Balage

**Comment by Flyer53** [comment]
Well, the code sample you posted above doesn't look like a hack to me. It's a very normal thing to check a condition and to continue based on the result. You could make the code a bit smaller though if you like:
```javascript
 jsPanel.layout.getDataset(this.panelId, 'id', this.storageName)
    ? jsPanel.layout.restoreId({id: this.panelId, config: options, storagename: this.storageName})
    : jsPanel.create(options);
```
The first argument in your previous post doesn't make a lot of sense to me because it's exactly the purpose of the layout extension to restore saved position and size. If you want to change those values don't use the layout extension's restore method and create the panel the usual way.

The second argument: I can't add extra code to the lib in order to prevent a user from making a mistake that's possible only because the developer forgot something. If I start doing this I'd have to add countless additions of code to the script. I don't think that's reasonable. And one of my goals always was, and still is, to keep the codebase as small as possible.

For the same reason argument 3 can't really be of my concern. Imagine how many additions I would have to make just to support testcases the developers might come up with. Again I think it's quite normal that a test environment for a new app has some extra custom code that's removed when the app is ready to be released.

Bottom line:
As far as I see it only the missing return value of the `restoreId()` method is a bug that needs a fix. The other points would bloat the codebase unreasonably I think.

**Comment by balage1551** [comment]
I accept your argument, although I don't see, why my request to make the whole CRUD life cycle of persistence within the API is a bad idea (usually an API which offers INSERT and UPDATE, also offers DELETE operation as well), but I completely honor your right as API designer to declare this out-of-scope.

I need this, so I will create my own extension and that's perfectly ok for me.

Thanks for fixing the `restoreId()` bug.

**Comment by Flyer53** [comment]
Well, I do understand you. But even the native JavaScript Storage API doesn't offer a method to update a stored item. It leaves you only the option to delete an item completely and set a new item instead. And remember that the Storage API stores only text values. Since the layout extension sets only one item there's no way to update this item. 

So far you are the very first user in need of a way to update a single "substring" within the stored value.

**Comment by balage1551** [comment]
Maybe that's my (sometimes bad) habit, but if I see an incomplete API, I often urge to create a wrapper to extend it. ;-)
I would wrap the JS Storage API with a tool to give some convenient methods, such as replacing part of the data, or removing it.

So forgive me! ;-) I think I can live without it. :-D

Also, I can use your ?: to restore or create the panel (as soon as your bugfix is available ;-)), but I think of it ("restore or create if not exists") is a widely used idiom, so it has its own right to get into the API. (If I'm not mistaken, the earlier version of your lib also done this: if restore failed, the panel created from the options.)

But you are completely right these are one-liners and it is only a taste where they are put: in lib or in my code.

**Comment by Flyer53** [comment]
Don't worry ... I always wellcome requests and new ideas, even if I can't implement everyone's wishes.
Stay healthy ... and happy coding,
Stefan
